Add support for selections to `st.plotly_chart`

willhuang1997 opened the pull request:
## Describe your changes

Adds selection support to `st.plotly_chart` that can be used via:

```python
selections = st.plotly_chart(fig, on_select="rerun", selection_mode=("box", "points"))
```

## GitHub Issue Link (if applicable)

- Related: https://github.com/streamlit/streamlit/issues/455
- Closes https://github.com/streamlit/streamlit/issues/5902
- Closes https://github.com/streamlit/streamlit/issues/8244
- Closes https://github.com/streamlit/streamlit/issues/8169
- Closes https://github.com/streamlit/streamlit/issues/7597
- Closes https://github.com/streamlit/streamlit/issues/6324
- Closes https://github.com/streamlit/streamlit/issues/8575
- Closes https://github.com/streamlit/streamlit/issues/8576

## Testing Plan

- Implemented e2e and unit tests.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

willhuang1997 committed:
Current prototype with plotly. it works with swapping pages now. need to investigate interaction with plotly and then interaction with another widget and then interaction with plotly bug

willhuang1997 committed:
Add hack for attribute dictionary for now

willhuang1997 committed:
Refactor code to store state in widgetStateManager so when deltaPath changes it retains state

willhuang1997 committed:
Change return value to box and lasso from select_box and select_lasso

willhuang1997 committed:
Fix small bug to always include box and lasso

willhuang1997 committed:
Add docstring to plotly chart

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Add python tests and change proto to be is_select_enabled

willhuang1997 committed:
Fix codeql errors

willhuang1997 committed:
Minor cleanup

willhuang1997 committed:
Add and fix react unit tests

willhuang1997 committed:
Minor fixes

willhuang1997 committed:
Make Attribute code more understandable

willhuang1997 committed:
Minor removeal of import statement

willhuang1997 committed:
Remove default orjson

willhuang1997 committed:
Add WidgetStateManager methods to store extra information

willhuang1997 committed:
Add make notices and upgrade plotly

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Fix merge conflict I messed up

willhuang1997 committed:
Minor cleanup

sfc-gh-wihuang commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -46,12 +43,86 @@ export interface PlotlyIFrameProps {
+  )
+}
+
+function keysToSnakeCase(obj: Record<string, any>): Record<string, any> {
```
This needs to be refactored to decamelize or done on the backend side.

willhuang1997 committed:
Refactor widgetStateManager methods

willhuang1997 committed:
Remove plotly-dist

willhuang1997 committed:
Cleanup attribute dictionary

willhuang1997 committed:
Remove print statement

willhuang1997 commented on lib/tests/streamlit/elements/plotly_chart_test.py:
```diff
@@ -107,3 +108,47 @@ def test_st_plotly_chart_sharing(self):
+
+    @parameterized.expand(
+        [
+            (True, True),
```
This should likely be changed as we will only be using `True / False` OR `"rerun"/"ignore"`, aka not both options.

willhuang1997 committed:
Minor cleanup

willhuang1997 committed:
Removing setSpec because it doesn't really seem to be doing anything...

willhuang1997 committed:
Minor cleanup

willhuang1997 committed:
Minor cleanup

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
fix jstest

willhuang1997 committed:
Fix lint error

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Fix plotly snapshot

willhuang1997 committed:
Fix Plotly State when double clicking when a selection is present (#8377)

<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes
- We send a widgetJson response of an empty dictionary when a double
click is issued. as a result, we need to check for that empty
dictionary. This is just adding that check
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?
  - manually tested
- Tests will be added in this PR:
https://github.com/streamlit/streamlit/pull/8255
---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

willhuang1997 committed:
Interactive charts plotly/e2e (#8432)

<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

willhuang1997 committed:
Use decamelize instead of helper function and other miscellaneous code changes (#8404)

<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes
- use decamelize
- fix regression for theming as useState would cache the spec value and
thus the theme value
- small fix for setting selectedPoints and selections to undefined when
we get an empty dictionary
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

willhuang1997 committed:
Fix mypy

willhuang1997 committed:
Merge branch 'feature/InteractiveCharts_Plotly' of https://github.com/streamlit/streamlit into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Fix bug where we return empty selection

willhuang1997 committed:
Fix lint

willhuang1997 committed:
Add new snapshots

willhuang1997 committed:
Add some wait time

willhuang1997 committed:
Add new snapshots

willhuang1997 committed:
Minor changes to reset callback

willhuang1997 committed:
Fix lint

lukasmasuch commented on lib/streamlit/runtime/state/common.py:
```diff
@@ -174,6 +177,8 @@ def failure(
     Message,
     PROTO_SCALAR_VALUE,
+    Data,
+    Dict[Any, Any],
```
I think we need to double-check if these are actually safe values. I actually don't think so since a dict just holds some references to other objects.

lukasmasuch commented on lib/streamlit/elements/plotly_chart.py:
```diff
@@ -78,8 +87,10 @@ def plotly_chart(
+        on_select: bool | Literal["rerun", "ignore"] | WidgetCallback = False,
         **kwargs: Any,
-    ) -> DeltaGenerator:
+    ) -> Union["DeltaGenerator", AttributeDictionary]:
```
nit: use the new annotation syntax here:

```suggestion
    ) -> "DeltaGenerator" | AttributeDictionary:
```

lukasmasuch replied:
Also, I think we should return a typed dict here instead of `AttributeDictionary`

lukasmasuch replied:
And, I think we can use signature overloading here as well. You can find how I did the signature overloading for dataframe selections here: https://github.com/streamlit/streamlit/pull/8411#diff-d7dbdd5a475ae010e6e559e2d749e249ec8b6a53cac3c23aa763d11d7b8a6bf7R127

I think we can do something very similar for Plotly here as well.

lukasmasuch commented on lib/streamlit/elements/plotly_chart.py:
```diff
@@ -148,20 +166,72 @@ def plotly_chart(
+        check_on_select_str(on_select, "plotly_chart")
+        if current_form_id(self.dg):
+            # TODO(willhuang1997): double check the message of this
+            raise StreamlitAPIException("st.plotly_chart cannot be used inside forms!")
```
This would break existing apps that use `plotly_charts` inside forms (which is possible right now). If you move this under a `if on_select != ignore` check as suggested below, it will likely be fine.

lukasmasuch replied:
Only run this logic `if on_select != ignore`, similar to how its done here: https://github.com/streamlit/streamlit/pull/8411#diff-d7dbdd5a475ae010e6e559e2d749e249ec8b6a53cac3c23aa763d11d7b8a6bf7R320

lukasmasuch commented on e2e_playwright/__snapshots__/linux/st_plotly_chart_select_test/st_plotly_chart-single_select[firefox].png:
Why is there no tooltip shown as in the other two screenshots?

lukasmasuch commented on frontend/lib/src/components/core/Block/ElementNodeRenderer.tsx:
```diff
@@ -412,6 +412,7 @@ const RawElementNodeRenderer = (
         <PlotlyChart
           element={node.element.plotlyChart as PlotlyChartProto}
           height={undefined}
+          widgetMgr={widgetProps.widgetMgr}
```
Instead of just adding the widgetMgr, it would be better to directly add the full widgetProps instead of elementProps below.

lukasmasuch replied:
Regarding the other widgetProps: We will likely also need to use `fragmentId` to add to the rerun requests so that it works with fragments. Also, I think we should disable selections in case the `disabled` property is true.

lukasmasuch commented on frontend/lib/src/util/utils.test.ts:
```diff
@@ -505,3 +506,22 @@ describe("getLoadingScreenType", () => {
 })
+
+describe("keysToSnakeCase", () => {
+  it("should remove . with _", () => {
```
Nit:
```suggestion
  it("should replace . with _", () => {
```

SimonBiggs commented:
Thank you oh so very much for this. Absolutely ❤️ this 🙂, thank you 🙂.

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

andfanilo commented:
Ahah same as Simon, I am VERY VERY VERY looking forward for this 😍

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Fix lint

willhuang1997 committed:
Fix lint again...

lukasmasuch commented on e2e_playwright/st_plotly_chart_select.py:
```diff
@@ -0,0 +1,191 @@
+fig_bubble.update_layout(dragmode="select")
+st.header("Bubble Chart with Box Select")
+st.plotly_chart(fig_bubble, on_select=True, key="bubble_chart")
+if st.session_state.bubble_chart:
```
Based on a discussion with @jrieke yesterday, the empty state of the data structure is still an open decision.   Something we need to wait for before merging this.

lukasmasuch commented on lib/streamlit/elements/plotly_chart.py:
```diff
@@ -78,8 +87,10 @@ def plotly_chart(
         sharing: SharingMode = "streamlit",
         theme: Literal["streamlit"] | None = "streamlit",
+        key: Key | None = None,
+        on_select: bool | Literal["rerun", "ignore"] | WidgetCallback = False,
```
This is also still an open decision if we support `bool` or `Literal["rerun", "ignore"]`. Something we need to wait for before merging this.

lukasmasuch commented on e2e_playwright/st_plotly_chart_select.py:
```diff
@@ -0,0 +1,191 @@
+else:
+    st.write("Nothing is selected")
+
+# TODO(willhuang1997): Readd choropleth charts
```
nit: is this still planned to be readded?

lukasmasuch replied:
ping. Either we fix this test or we remove it from here

lukasmasuch commented on e2e_playwright/st_plotly_chart_select_test.py:
```diff
@@ -0,0 +1,135 @@
+
+
+def test_box_select_on_stacked_bar_chart_displays_a_df(app: Page):
+    chart = app.locator(".stPlotlyChart").nth(3)
```
not 100% sure on this, but maybe we should just wrap the plotly chart into a div to be able to add a `stPlotlyChart` test ID. I think this would make it a lot more consistent and the overhead with the div would be reasonable.

sfc-gh-wihuang replied:
I think it's not a good idea to do this because sometimes the div will load but the actual plotly chart wouldn't have finished loading.

lukasmasuch commented on frontend/lib/src/WidgetStateManager.ts:
Probably migrate to this a bit more generic version for element state: https://github.com/streamlit/streamlit/pull/8468

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -46,12 +44,61 @@ export interface PlotlyIFrameProps {
+      y: [],
+    }
+  }
+  const points = pathData.replace("M", "").replace("Z", "").split("L")
```
nit: can you add a quick comment here with some description on how the `pathData` format looks like?

lukasmasuch commented on frontend/lib/src/util/utils.ts:
```diff
@@ -500,3 +501,27 @@ export function extractPageNameFromPathName(
   )
 }
+
+export function keysToSnakeCase(
```
nit: Can you add a short function docstring to add some additional describe for this utils method?

lukasmasuch commented on lib/streamlit/attribute_dictionary.py:
```diff
@@ -0,0 +1,27 @@
+        except KeyError:
+            raise AttributeError(
+                f"'{type(self).__name__}' object has no attribute '{key}'"
+            )
```
Can we overwrite this to auto-convert all normal dics to attribute dicts? E.g.:

```suggestion
    def __getattr__(self, key):
        try:
            item = self.__getitem__(key)
            if isinstance(item, dict):
               # If the item is a dict, wrap it again into an `AttributeDictionary` to allow dot notation access. 
               return AttributeDictionary(item)
             return item
        except KeyError:
            raise AttributeError(
                f"'{type(self).__name__}' object has no attribute '{key}'"
            )
```

sfc-gh-wihuang replied:
Done

lukasmasuch commented on lib/streamlit/attribute_dictionary.py:
```diff
@@ -0,0 +1,27 @@
+from typing import Any, Dict
+
+
+class AttributeDictionary(Dict[Any, Any]):
```
nit: Can you add a short docstring here describing what this `AttributeDictionary` can be us for / or what it does

sfc-gh-wihuang replied:
Done

lukasmasuch commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -70,6 +70,7 @@
         "time_input": "string_value",
         "component_instance": "json_value",
         "data_editor": "string_value",
+        "plotly_chart": "json_value",
```
I would recommend to actually use the `string_value` here as well. The `json_value` doesn't really simplify anything, it makes aspects even a bit more complicated. Thats why I decided to use `string_value` for `data_editor` as well as the dataframe selections here: https://github.com/streamlit/streamlit/pull/8411

You can also find the `DataframeSelectionSerde` implementation here: https://github.com/streamlit/streamlit/pull/8411#diff-d7dbdd5a475ae010e6e559e2d749e249ec8b6a53cac3c23aa763d11d7b8a6bf7R101 which you probably can directly adapt.

sfc-gh-wihuang replied:
Done. Thanks for linking the code.

lukasmasuch commented on lib/streamlit/elements/__init__.py:
```diff
@@ -32,6 +32,7 @@
     "text_area",
     "time_input",
     "date_input",
+    "plotly_chart",
```
Moving this to `WIDGETS` will likely break all apps that use plotly within caching functions :( We probably need to refactor the way on how we do the widget_replay check. I will take a closer look on that.

lukasmasuch commented on lib/streamlit/elements/plotly_chart.py:
```diff
@@ -23,13 +23,22 @@
 from streamlit.elements.lib.streamlit_plotly_theme import (
     configure_streamlit_plotly_theme,
 )
+from streamlit.elements.utils import check_callback_rules, check_session_state_rules
```
Not sure why this cyclic import warning happens. But I decided to move the imports down into the function. See here: https://github.com/streamlit/streamlit/pull/8411#diff-d7dbdd5a475ae010e6e559e2d749e249ec8b6a53cac3c23aa763d11d7b8a6bf7R322

lukasmasuch commented on proto/streamlit/proto/PlotlyChart.proto:
```diff
@@ -36,6 +36,10 @@ message PlotlyChart {
 
+  string id = 7;
+
+  bool is_select_enabled = 8;
```
Based on a discussion with @jrieke yesterday, we might consider also adding a `selection_mode` to plotly that allows to select the events that should trigger reruns. 

`selection_mode = "box"` would 1) remove the `lasso` selection toolbar item 2) only trigger the rerun if there is actually a change in with the `box` selection state. 

This is quite similar to how we might do this with dataframe selections (columns & rows): https://github.com/streamlit/streamlit/pull/8411

If we do that, we likely would need to change this to an enum like it is done here: https://github.com/streamlit/streamlit/pull/8411/files#diff-d585065d5d6302061bb168e8ad4523d034388edca3390075dd27fbbe2784f652R46

sfc-gh-wihuang replied:
Should that be done in a fast follow PR?

lukasmasuch replied:
This could be done partly in a follow up, but some aspects would need to be already adapted to that (e.g., the proto).

lukasmasuch commented on lib/streamlit/elements/plotly_chart.py:
```diff
@@ -148,20 +166,72 @@ def plotly_chart(
+
+        self.dg._enqueue("plotly_chart", plotly_chart_proto)
+        if is_select_enabled:
+            return AttributeDictionary(widget_state.value)
```
We probably should run all of this logic only if `on_select` is actually activated otherwise, we will add some overhead to all uses that actually don't need the selections. Also see how this is done in the dataframe selections PR here: https://github.com/streamlit/streamlit/pull/8411/files#diff-d7dbdd5a475ae010e6e559e2d749e249ec8b6a53cac3c23aa763d11d7b8a6bf7R381-R425

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
-    } else {
-      spec.layout.width = initialWidth
-      spec.layout.height = initialHeight
+    const storedValue = widgetMgr.getJsonValue(element)
```
nit: as suggested above, I would probably suggest changing this to `string` here. The `json_value` is mostly an artifact of the past and adds some complexity.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
-      spec.layout.height = initialHeight
+    const storedValue = widgetMgr.getJsonValue(element)
+
+    if (storedValue === "{}") {
```
nit: I think we should parse this to json first, and then do the empty check on the parse json structure.

lukasmasuch replied:
Also, `storedValue` can also be undefined or? Don't we have to check this here as well?

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+          element,
+          SELECTIONS_KEY
+        )
+        spec.data = data
```
nit: can you add a comment here what is actually contained/expected in `spec.data`  and `spec.layout.selections`.
E.g. is  `spec.data` all data of the chart, or only the selected one? If it is all data, can't we just take it from the `element` instead of storing it in the state?

lukasmasuch replied:
Is there a link you can add here to the plotly side with more details on the spec format?

sfc-gh-wihuang replied:
`spec.data` = `data` + `selectionpoints` from our `onSelect` event. Will add a link.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+          // make all other points opaque
+          spec.data.forEach((trace: any) => {
+            if (!trace.selectedpoints) {
+              trace.selectedpoints = []
```
What is the reason here why it gets to be set `opaque` when doing that. Is it because: `trace.selectedpoints === null || trace.selectedpoints === undefined` would show all points. But if its actually empty, plotly greys out the points? Maybe you can add more explanation here.

sfc-gh-wihuang replied:
Added a comment here. 
If it's empty, it will grey out the points. If it has some values, then it will highlight / keep their 100% opaqueness.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
-  }, [element, theme, height, width, figure.config, generateSpec])
+  }, [element, figure.spec, theme, widgetMgr])
+
+  const spec = getInitialValue()
```
We might want to put this into a `useMemo` to cache this value, assuming that we don't want to re-execute `getInitialValue` on every single rendering.

sfc-gh-wihuang replied:
We do want to re-execute `getInitialValue` on every rerender, otherwise the graph will be out of sync as far as I can tell when testing.

I was using `useState` to cache the value but then the graph would not rerender properly.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+
+  const spec = getInitialValue()
+
+  const [initialHeight] = useState(spec.layout.height)
```
Can you add a comment here why it needs to be wrapped in a `useState`? Is it because we try to reconstruct the spec from the state which will contain adapted versions of the spec?

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+  const [initialHeight] = useState(spec.layout.height)
+  const [initialWidth] = useState(spec.layout.width)
+
+  if (isFullScreen(height)) {
```
There is actually an `isFullScreen` flag that gets passed to all fullscreen components. This is preferred to checking this via the height. See Dataframe.tsx for example: 
https://github.com/streamlit/streamlit/blob/55bba786dcc8dd85e724746cd26bba3018a147bc/frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx#L120

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+    spec.layout.height = height
+  } else if (element.useContainerWidth) {
+    spec.layout.width = width
+    if (!isFullScreen(height) && height !== initialHeight) {
```
I think the `!isFullScreen(height)` is unnecessary here since if it would be in fullscreen mode, the if condition above would have been chosen:
```suggestion
    if (height !== initialHeight) {
```

lukasmasuch replied:
Also not sure if the check makes sense since, `height` would always be `undefined` if `!isFullScreen(height)` or? So the check is actually: undefined !== initialHeight. Why is this needed in the first place?

sfc-gh-wihuang replied:
This is bad leftover prototype code and the fullscreen wasn't properly being rendered so I think I was playing with it. Will remove it.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+    spec.layout.height = initialHeight
+  }
+  if (element.isSelectEnabled) {
+    spec.layout.clickmode = "event+select"
```
nit: Can you add a one-liner comment to say why we set `clickmode` and why we set `hovermode` here

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+    returnValue.select.points = returnValue.select.points.map((point: any) =>
+      keysToSnakeCase(point)
+    )
+    widgetMgr.setJsonValue(element, returnValue, { fromUi: true })
```
I think we need to pass in the `fragmentId` here as well so that it works with fragments.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
 
+  const reset = useCallback((): void => {
+    widgetMgr.setExtraWidgetInfo(element, SELECTIONS_KEY, {})
+    widgetMgr.setJsonValue(element, {}, { fromUi: true })
```
I think we need to pass in the `fragmentId` here as well so that it works with fragments.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
       layout={layout}
       config={config}
       frames={frames}
+      onSelected={element.isSelectEnabled ? handleSelect : () => {}}
```
Do you know what the difference is between `onSelected` and `onSelecting` is based on the callback list [here](https://github.com/plotly/react-plotly.js/tree/24d81259ea9369074586f1c3d2b7ae9fc3610860?tab=readme-ov-file#event-handler-props)?

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
     <Plot
       key={isFullScreen(height) ? "fullscreen" : "original"}
       className="stPlotlyChart"
+      divId={element.id}
```
Just curious: is this needed to be added? Even if not required, it might not hurt to do that. Just something to double check since we are not setting the div id for most other widgets.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+      event.selections.length === 0
+    ) {
+      return
+    }
```
nit: Instead of having to add all the `// @ts-expect-error` in the code, its probably better to just put this into a local variable:

```suggestion
    // event.selections doesn't show up in the PlotSelectionEvent
    // @ts-expect-error
    const selections = event.selections
    
    if (
      event.points.length === 0 &&
     selections  && 
     selections.length === 0
    ) {
      return
    }
```

lukasmasuch replied:
But I'm not sure if the check is full correct: If `event.selections` is an empty list, it wouldn't even go into the if because `event.selections` would be false or? You might actually want to do a `notNullOrUndefined(event.selections)` check here instead or?

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+    returnValue.select.box = selectedBoxes
+    returnValue.select.lasso = selectedLassos
+    returnValue.select.points = returnValue.select.points.map((point: any) =>
+      keysToSnakeCase(point)
```
The `keysToSnakeCase` does go through the full nested data structure or? Wouldn't it be possible to just apply it on the foll `returnValue` or `returnValue.select` dict?

sfc-gh-wihuang replied:
box and lasso should already be converted to the correct casing with `parseLassoPath` and `parseBoxSelection` so it's unnecessary to do it for the `returnValue` as far as I can tell.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+        legendgroup: point.data.legendgroup
+          ? point.data.legendgroup
+          : undefined,
+        data: undefined,
```
We are cutting this out because it adds unnecessary data overhead or? Can you add a comment here with the reason why we cut this out.

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+    event.points.forEach(function (point: any) {
+      selectedPoints.push({
+        ...point,
+        legendgroup: point.data.legendgroup
+          ? point.data.legendgroup
+          : undefined,
```
nit: Can be simplified to:

```suggestion
        legendgroup: point.data.legendgroup || undefined
```

lukasmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -73,65 +120,194 @@ function PlotlyFigure({
+    }
+
+    event.points.forEach(function (point: any) {
+      selectedPoints.push({
+        ...point,
+        legendgroup: point.data.legendgroup
+          ? point.data.legendgroup
+          : undefined,
+        data: undefined,
+        fullData: undefined,
+      })
+      pointIndices.push(point.pointIndex)
+
+      // build graph representation to retain state
+      if (
+        data[point.curveNumber] &&
+        !data[point.curveNumber].selectedpoints.includes(point.pointIndex)
+      ) {
+        data[point.curveNumber].selectedpoints.push(point.pointIndex)
+      } else {
+        data[point.curveNumber].selectedpoints = [point.pointIndex]
+      }
+    })
```
I'm not sure if this code is fully correct. Is it expected that: `data[point.curveNumber].selectedpoints = [point.pointIndex]` will get executed if either `data[point.curveNumber]` is falsy or if `point.pointIndex` is not included in `data[point.curveNumber].selectedpoints`?

sfc-gh-wihuang replied:
Hm what I want to do here is add each `pointIndex` to the corresponding trace (`data[point.curveNumber]`). 

I attempt to add it if an array already exists. 
If it doesn't initialize the array with the pointIndex. 

Maybe I should check that `data[point.curveNumber]` is not null and undefined rather than falsy?

lukasmasuch commented on lib/streamlit/chart_util.py:
I would probably for now, move this logic back into the plotly method and revisit sharing some code if we have some more selections implemented across our code base. Since we likely also share similar selection logic with other non-chart elements it might be better to have it shared in a `selection_util`. But for now, just for this short check its doesn't seem worth it to creat a new util module yet.

sfc-gh-wihuang replied:
I think this will be used in st.line_chart, st.vega_lite_chart, st_altair_chart so I think it's worth to keep here. Thoughts?

lukasmasuch commented on e2e_playwright/__snapshots__/linux/st_plotly_chart_select_test/st_plotly_chart-single_select[firefox].png:
Why is there no tooltip shown as in the other two screenshots?

willhuang1997 committed:
fix nits

willhuang1997 committed:
Add docstring for keyToSnakeCase

willhuang1997 committed:
Fix codeql and github action checks

willhuang1997 committed:
fix lint

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Add firefox snapshots?

willhuang1997 committed:
Remove snapshots and change plotly chart to have plotly toolbar

willhuang1997 committed:
Remove orjson line

willhuang1997 committed:
Submit dark version of plotly chart that is constantly failing

willhuang1997 committed:
Add snapshot with firefox having the plotly toolbar

willhuang1997 committed:
looks like locally displayModeBar being false works so trying to readd it

willhuang1997 committed:
Add new snapshots

willhuang1997 committed:
Merge branch 'playwright/fixPlotlyPlaywrightSnapshots' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Fix snapshots for some reason that are failing

willhuang1997 committed:
Fix nits and migrate to setStringValue rather than JsonValue for frontend and backend

sfc-gh-wihuang commented on e2e_playwright/__snapshots__/linux/st_plotly_chart_test/st_plotly_chart-line-chart-specific-height-width[light_theme-webkit].png:
Looks like the main difference is the buttons moved over which I think is fine? Not sure of root cause why.

willhuang1997 committed:
Remove print statements

willhuang1997 committed:
Fix e2e test as the return value changed by using string value

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Refactor with new ElementState WidgetStateManager methods

willhuang1997 committed:
Remove down statements in e2e test

willhuang1997 committed:
Remove one more down that I forgot...

willhuang1997 committed:
Add docstring

willhuang1997 committed:
fix bug that allowed selections to enter non-select plotly charts

lukasmasuch committed:
Evaluate different mechanism to keep plotly state (#8515)

## Describe your changes

Use a different mechanism using `onUpdated` callback to keep the Plotly
state.

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

---------

Co-authored-by: William Wei Huang <william.huang@snowflake.com>

willhuang1997 committed:
Attempt to remove bool for on select param (#8525)

<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes
- refactor removing boolean because "rerun" and "ignore" are the
preferred ways of setting `on_select`
- should likely wait to get this PR in until other PRs such as
https://github.com/streamlit/streamlit/pull/8522 and
https://github.com/streamlit/streamlit/pull/8515 are merged
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

---------

Co-authored-by: lukasmasuch <lukas.masuch@gmail.com>

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/InteractiveCharts_Plotly

lukasmasuch committed:
Merge branches 'feature/InteractiveCharts_Plotly' and 'feature/InteractiveCharts_Plotly' of https://github.com/streamlit/streamlit into feature/InteractiveCharts_Plotly

lukasmasuch committed:
Add signature overloading

willhuang1997 committed:
Fix id generation for plotly and remove small comment

lukasmasuch committed:
Apply more refactorings and improvements

lukasmasuch committed:
Merge branch 'feature/InteractiveCharts_Plotly' of https://github.com/streamlit/streamlit into feature/InteractiveCharts_Plotly

willhuang1997 commented on proto/streamlit/proto/PlotlyChart.proto:
```diff
@@ -36,7 +36,22 @@ message PlotlyChart {
+  string id = 7;
+
+  // Activate selections types on the chart.
+  repeated SelectionMode selection_mode = 12;
```
I think this should be 8? I assume this is copied and pasted from Dataframe selections PR

lukasmasuch replied:
oh yes, good catch

lukasmasuch committed:
Improve reset handling

lukasmasuch committed:
Add some more checks

lukasmasuch committed:
Add more logs

lukasmasuch committed:
Fix issue

lukasmasuch committed:
Change back to plotly mode

lukasmasuch committed:
Change default to ignore

lukasmasuch committed:
Correctly handle initial state

lukasmasuch committed:
Always respect available width as maximum

lukasmasuch committed:
Fix fragment support

lukasmasuch committed:
Migrate fullscreen button to plotly toolbar

lukasmasuch committed:
Further fixes

lukasmasuch committed:
Don't update when selection mode is not activated

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/InteractiveCharts_Plotly

lukasmasuch committed:
Clean up safe types

lukasmasuch committed:
Minor refactorings

lukasmasuch committed:
More refactorings

lukasmasuch committed:
Add more dependencies

lukasmasuch committed:
Remove not needed code

lukasmasuch committed:
Some refactoring into functions

lukasmasuch committed:
Deprecate the plotly sharing mode

lukasmasuch committed:
Minor refactoring

lukasmasuch committed:
Apply more improvements

lukasmasuch committed:
Remove unnecessary check

lukasmasuch committed:
More refactoring

lukasmasuch committed:
Remove some comments

willhuang1997 committed:
Fix typechecking for Interactive Plotly Charts (#8550)

## Describe your changes
- best attempt to fix typechecking for plotly
- there is one # type: ignore at line 185 because mypy is not
recognizing that they're compatible sets
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

lukasmasuch committed:
Refactor types

lukasmasuch committed:
Remove unused interface

lukasmasuch committed:
Remove more console logs

lukasmasuch committed:
Add tracing

lukasmasuch committed:
Revert

lukasmasuch committed:
Remove todo comment

lukasmasuch committed:
Remove console log

lukasmasuch committed:
Small move

lukasmasuch committed:
More cleanup

willhuang1997 committed:
Fix Python tests for Interactive Plotly Charts (#8551)

## Describe your changes
- fix python tests
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

---------

Co-authored-by: Lukas Masuch <Lukas.Masuch@gmail.com>

willhuang1997 committed:
Fix Playwright tests for Interactive Plotly Charts (#8552)

## Describe your changes
- new fullscreen changes have come in so need to fix the playwright code
to look for new full screen / exit full screen buttons
- new fullscreen changes so snapshots likely need to be regenerated
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

---------

Co-authored-by: Lukas Masuch <Lukas.Masuch@gmail.com>

lukasmasuch committed:
Use Widget Callback

willhuang1997 committed:
Remove default height

willhuang1997 committed:
Merge branch 'feature/InteractiveCharts_Plotly' of https://github.com/streamlit/streamlit into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Remove Callable from typing as it's unused

willhuang1997 committed:
Add new updated snapshots

willhuang1997 committed:
Fix typing

willhuang1997 committed:
Remove junit tests because it's covered in e2e tests

willhuang1997 committed:
Add extra check for removing plotly config buttons

willhuang1997 committed:
Fix lint

willhuang1997 committed:
Add snapshots

willhuang1997 committed:
Fix failing snapshots

willhuang1997 committed:
Fix snapshots

willhuang1997 committed:
Add Interactive Plotly typescript unit tests (#8557)

## Describe your changes
- title
- added tests for:
  - `sendEmptySelection`
  - `handleSelection`
  - `applyTheming`
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

willhuang1997 committed:
Fix double click Plotly behavior (#8561)

## Describe your changes
- null was being passed to resetSelectionInFigure because Plotly will
pass in null. As a result, we need to actually call the function ourself
to get the default value
- doing this also changes the default plotly pan behavior and also
resets selections too
Current:


https://github.com/streamlit/streamlit/assets/16749069/cb14a994-7d0f-4425-9c4f-a660ae9248bf


Now:


https://github.com/streamlit/streamlit/assets/16749069/8820f163-4294-4781-b12d-04d6a750ab12


## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

willhuang1997 committed:
Fix comment and remove some other comments

sfc-gh-lmasuch commented on frontend/lib/src/components/elements/PlotlyChart/PlotlyChart.tsx:
```diff
@@ -16,150 +16,724 @@
+      ]
+    }
+
+    if (!config.modeBarButtonsToRemove && isSelectionActivated) {
```
nit: we do want to apply these toolbar modifications in all cases (basically implementing what is suggested in this [project](https://www.notion.so/snowflake-corp/Rethink-Plotly-toolbar-f0b19b1e31494d5a9e3d224e2b93e301)). So its correct to apply this always:

```suggestion
    if (!config.modeBarButtonsToRemove) {
```

sfc-gh-lmasuch replied:
also see the part below:

```python
if (!isSelectionActivated) {
        // Remove lasso & select buttons in read-only charts:
        modeBarButtonsToRemove.push("lasso2d", "select2d")
 }
```

willhuang1997 committed:
Fix point selection being enabled when it shouldnt be

sfc-gh-wihuang committed:
Add tests for selection_mode and fix bug where single selection mode wasn't working (#8570)

willhuang1997 committed:
Revert "Fix point selection being enabled when it shouldnt be"

This reverts commit 66983cec27b7b322dacdb2631311292d540f3739.

lukasmasuch committed:
Add unit test for widget replay

lukasmasuch committed:
More unit tests

lukasmasuch committed:
Add another test

lukasmasuch committed:
Improve test

lukasmasuch committed:
Deactivate point selection if selection mode doesn't include points (#8574)

## Describe your changes

## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

---------

Co-authored-by: willhuang1997 <willhuang1997@gmail.com>

sfc-gh-wihuang committed:
Remove unnecessary buttons in Plotly toolbar (#8571)

## Describe your changes
- removes config buttons even for non-selection plotly charts
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

---------

Co-authored-by: willhuang1997 <willhuang1997@gmail.com>

lukasmasuch committed:
Only activate selection types if selections are activated

lukasmasuch committed:
Use next line eslint ignores for dependencies

lukasmasuch commented on e2e_playwright/st_plotly_chart_select.py:
```diff
@@ -0,0 +1,189 @@
+    log_x=True,
+    size_max=60,
+)
+fig_bubble.update_layout(dragmode="select")
```
I think we should remove most of the update_layout dragmode in this test since it is anyways determined automatically and only add it maybe to one test (e.g. by setting a box selection to dragmode pan) to make sure that the user can overwrite this.

```suggestion
```

lukasmasuch commented on e2e_playwright/st_plotly_chart_select.py:
```diff
@@ -0,0 +1,189 @@
+    wide_df, x="nation", y=["gold", "silver", "bronze"], title="Wide-Form Input"
+)
+fig.update_layout(dragmode="select")
+st.plotly_chart(
```
All of these tests set the `key` and use the session state value. For better test coverage, it would be good to have one of these tests use the return value instead + not set the key.

lukasmasuch commented on e2e_playwright/st_plotly_chart_select.py:
```diff
@@ -0,0 +1,189 @@
+
+
+def histogram_callback():
+    try:
```
Instead of try & except here, it would probably be better to use the same if check as in the tests above.

lukasmasuch commented on e2e_playwright/st_plotly_chart_select_test.py:
```diff
@@ -0,0 +1,182 @@
+    app.mouse.down()
+    app.mouse.move(150, 150)
+    app.mouse.up()
+    wait_for_app_run(app, 3000)
```
nit: I don't think we need to wait an additional 3 seconds here since the next operation is a waiting except operation. I think this timeout might be only relevant if there is an `assert_snapshot`that follows directly.

LukasMasuch commented on e2e_playwright/st_plotly_chart_select_test.py:
```diff
@@ -0,0 +1,182 @@
+    app.mouse.move(400, 400)
+    app.mouse.move(435, 500)
+    app.mouse.up()
+    wait_for_app_run(app, 3000)
```
nit: I don't think we need to wait an addition 3 seconds here since the next operation is anyways a waiting operation

LukasMasuch commented on e2e_playwright/st_plotly_chart_select_test.py:
```diff
@@ -0,0 +1,182 @@
+    chart.hover()
+    app.mouse.down()
+    app.mouse.up()
+    wait_for_app_run(app, 3000)
```
nit: I don't think we need to wait an additional 3 seconds here since the next operation is anyways a waiting operation

LukasMasuch commented on e2e_playwright/st_plotly_chart_select_test.py:
```diff
@@ -0,0 +1,182 @@
+    chart.scroll_into_view_if_needed()
+    chart.hover()
+    app.mouse.dblclick(400, 400)
+    wait_for_app_run(app, 3000)
```
nit: I don't think we need to wait an additional 3 seconds here since the next operation is a waiting except operation

LukasMasuch commented on e2e_playwright/st_plotly_chart_select_test.py:
```diff
@@ -0,0 +1,182 @@
+    app.mouse.down()
+    app.mouse.move(150, 150)
+    app.mouse.up()
+    wait_for_app_run(app, 3000)
```
nit: I don't think we need to wait an additional 3 seconds here since the next operation is a waiting except operation

LukasMasuch commented on e2e_playwright/st_plotly_chart_select_test.py:
```diff
@@ -0,0 +1,182 @@
+    app.mouse.move(400, 400)
+    app.mouse.move(435, 500)
+    app.mouse.up()
+    wait_for_app_run(app, 3000)
```
nit: I don't think we need to wait an additional 3 seconds here since the next operation is a waiting except operation

LukasMasuch commented on e2e_playwright/st_plotly_chart_select_test.py:
```diff
@@ -0,0 +1,182 @@
+    app.mouse.down()
+    app.mouse.move(150, 150)
+    app.mouse.up()
+    wait_for_app_run(app, 3000)
```
nit: I don't think we need to wait an additional 3 seconds here since the next operation is a waiting except operation

LukasMasuch commented on e2e_playwright/st_plotly_chart_select_test.py:
```diff
@@ -0,0 +1,182 @@
+    app.mouse.down()
+    app.mouse.move(150, 150)
+    app.mouse.up()
+    wait_for_app_run(app, 3000)
```
nit: I don't think we need to wait an additional 3 seconds here since the next operation is a waiting except operation

lukasmasuch commented on e2e_playwright/st_plotly_chart_select.py:
```diff
@@ -0,0 +1,189 @@
+
+
+fig.update_layout(dragmode="lasso")
+st.plotly_chart(
```
Is there any e2e test that actually tests the callback case?

willhuang1997 replied:
Hm how I understand it is:
the callback is showing a dataframe so if it's showing a dataframe and we expect one, then we would be testing the callback case. Am I missing something?

lukasmasuch committed:
Fix selection mode parsing

lukasmasuch committed:
Improve selection mode parsing test

lukasmasuch commented:
@willhuang1997 can you add an additional e2e test to check if the state of an component is correctly kept if its unmounted & remounted? E.g.:

```python
import time
if st.button("Create some elements to unmount component"):
    for i in range(3):
        # The sleep here is needed, because it won't unmount the
        # component if this is too fast.
        time.sleep(1)
        st.write("Another element")

st.plotly_chart(..., on_select="rerun")
```

And in the test, you do a box selection, then click on the button, and after the rerun finished, check if the selection is still valid (dataframe is on screen) + make a screenshot of the chart to check if the box selection is still on the chart.

lukasmasuch commented:
Another e2e test that I believe is not yet covered is a chart that allows multiple selections (e.g., points & box) and where you test that both selections can be used.

lukasmasuch committed:
Recreate snapshots

lukasmasuch committed:
Fix plotly tab change flickering (#8577)

## Describe your changes

Improve the flickering issue when the width of the vertical block isn't
determined yet (==-1).

## GitHub Issue Link (if applicable)

- Related: https://github.com/streamlit/streamlit/issues/8575

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/InteractiveCharts_Plotly

lukasmasuch committed:
Add updated snapshots

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/InteractiveCharts_Plotly

lukasmasuch committed:
Some cleanups

willhuang1997 committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 commented:
> Another e2e test that I believe is not yet covered is a chart that allows multiple selections (e.g., points & box) and where you test that both selections can be used.

@LukasMasuch , I tried doing a double selection but it looks like for some reason, playwright isn't accepting the "Shift" key and it won't even swap to lasso select... 

Should we just ignore the multiple selection as an e2e test and add a unit test?

branch for reference: https://github.com/streamlit/streamlit/pull/8579 that has the commit: 3bc01e40c4f2a4702fdeda150119665e989bbc60

https://github.com/streamlit/streamlit/assets/16749069/1341acfd-8004-42fc-b0cb-e22604e771d9

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Fix Interactive Plotly e2e playwright nits (#8579)

## Describe your changes
- add new tests
  - unmounting + mounting
  - choropleth chart (readded and uncommented out)
- fix some minor nits such as removing updating layout and removing
session_state checks
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

---------

Co-authored-by: Lukas Masuch <Lukas.Masuch@gmail.com>

lukasmasuch committed:
Mini changes

lukasmasuch committed:
Minor fine-tuning

lukasmasuch committed:
More improvements

lukasmasuch committed:
Some more tweaks

lukasmasuch committed:
More tweaks

lukasmasuch committed:
Remove test

lukasmasuch committed:
More fine-tuning

lukasmasuch committed:
Click somewhere else

lukasmasuch committed:
More fine-tuning

lukasmasuch committed:
Update snapshots

lukasmasuch committed:
Add test with box & points

lukasmasuch committed:
More changes

lukasmasuch committed:
Just check point count and not dataframe visuals

lukasmasuch committed:
Minor proto restructuring

lukasmasuch committed:
Remove flakiness

lukasmasuch committed:
Update snapshots

lukasmasuch committed:
Merge branch 'develop' into feature/InteractiveCharts_Plotly

willhuang1997 committed:
Add unit test for box and lasso selection  (#8585)

## Describe your changes
- add unit test for box AND lasso selection with mock data
## GitHub Issue Link (if applicable)

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

willhuang1997 committed:
Remove follow redirects in yarn lock

lukasmasuch committed:
Add hover before snapshots

willhuang1997 committed:
Add new snapshots

willhuang1997 committed:
Add new snapshots

willhuang1997 committed:
Add new updated snapshots

lukasmasuch committed:
Add test for deprecation warning

lukasmasuch committed:
Improve docstring

lukasmasuch reviewed (approved):
LGTM 📈🎉

willhuang1997 committed:
Remove extra quotations"

willhuang1997 committed:
add final new snapshot

lukasmasuch committed:
Remove choropleth chart

lukasmasuch committed:
Remove unused dependencies

lukasmasuch committed:
Update snapshots

lukasmasuch committed:
Improve test

willhuang1997 committed:
Fix test

willhuang1997 committed:
Fix test

lukasmasuch committed:
Use direct chart click

willhuang1997 committed:
Add new screenshots(?)

ppeetteerrs commented:
It would be great if we can limit the selection type to only x or y axis too (instead of box)
